docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,592 @@
+> It is useful to provide an audit trail of the changes to the fleet templates and sources.
+> It's also useful to enable implementations where offline devices can traverse a
+> history of configurations to reach the current state.
+
+When `fleet.spec.template` is updated the Fleet deployment changes, a copy of the device template
+is stored as a `TemplateVersion` object. This object freezes a copy of the template and also creates
+a list of external references (like git refs, image tags, etc..) that are floating tags by scanning
+all fleet devices, these floating tags are resolved to stable hashes and stored in (or along)
+the `TemplateVersion` object.
+
+When a previously frozen source changes, a new copy of `TemplateVersion` object is created
+with the updated source reference.
+
+If an error happens during the creation of the `TemplateVersion` object, the fleet controller will
+mark a status condition on the fleet object (e.g. missing sources, missing floating tags, etc..),
+and the fleet controller will not proceed with the rollout.
+
+Once a `TemplateVersion` is created the fleet controller starts rolling out that update to devices.
+This update can take a long time (depending on the upgrade strategy). The fleet controller marks
+the latest template version that is being rolled out, as a `fleet-controller/templateVersion`
+annotation in the fleet for observability through the API.
+
+
+When a template is applied to a specific device via the Device API:
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes),
+  this gives flightctl the ability to check images being deployed on devices before the device could auto pick-up
+  a floating tag. (i.e. checks verifying the installation of the flightctl agent, and configuration)
+* When later in time a device has labels updated, the device controller will queue the device to be re-evaulated,
+  for example if a change of fleet owner happens because of the label changes.
+* An `annotation` is added to the device in the form of `fleet-controller/templateVersion` to track the latest device
+  template applied to the device.
+
+TemplateVersions are exposed via the API as read-only objects, they can be used to track changes to the fleet
+spec and the external references that were frozen on the devices:
+
+* `/fleet/{name}/templateversions` - This endpoint enables listing of `TemplateVersion` objects for the fleet.
+* `/fleet/{name}/templateversions/{version}` - This endpoint returns a specific `TemplateVersion` object for the fleet.
+
+#### Device management within a fleet
+
+##### Removing a device from a fleet
+
+A device could be stripped from the labels associating it to a specific fleet, in such
+case the `metadata.owner` field will be cleared by the fleet controller, and the device
+will need to be manually manage or eventually associated to a different fleet/owner.
```

Comment:
How will we handle the potential case of certs expiring while the device is unmanaged? Or will the device even if unmanaged always be able to send a CSR to the flightctrl service. Not sure if we need to cover that at all here.

Reply:
Yes, a device being connected to the system means that we will renew the certificates, it's an orthogonal process as @avishayt said.

Then... while the device is managed by the system... it could be individual, an not managed by a fleet.

Reply:
I am making it clear by proposing a small change:

the device **spec** will need to be manually managed.